OS X 10.10: /usr/include/dispatch/object.h doesn't work with gcc

This is hit by many packages, and manifests as follows:
```
[gettext-bootstrap] libtool: compile:  gcc -c -DLOCALEDIR=\"/Users/certik/.hashdist/bld/gettext-bootstrap/spt6nwddjz7d/share/locale\" -DLOCALE_ALIAS_PATH=\"/Users/certik/.hashdist/bld/gettext-bootstrap/spt6nwddjz7d/share/locale\" -DLIBDIR=\"/Users/certik/.hashdist/bld/gettext-bootstrap/spt6nwddjz7d/lib\" -DBUILDING_LIBINTL -DBUILDING_DLL -DIN_LIBINTL -DENABLE_RELOCATABLE=1 -DIN_LIBRARY -DINSTALLDIR=\"/Users/certik/.hashdist/bld/gettext-bootstrap/spt6nwddjz7d/lib\" -DNO_XMALLOC -Dset_relocation_prefix=libintl_set_relocation_prefix -Drelocate=libintl_relocate -DDEPENDS_ON_LIBICONV=1 -DHAVE_CONFIG_H -I. -I. -I.. -I/Users/certik/.hashdist/bld/libiconv-bootstrap/y6dz7hsnygn3/include -g -O2 -fvisibility=hidden ./log.c  -fno-common -DPIC -o .libs/log.o
[gettext-bootstrap] In file included from /usr/include/dispatch/dispatch.h:51:0,
[gettext-bootstrap]                  from /System/Library/Frameworks/CoreFoundation.framework/Headers/CFStream.h:15,
[gettext-bootstrap]                  from /System/Library/Frameworks/CoreFoundation.framework/Headers/CFPropertyList.h:13,
[gettext-bootstrap]                  from ./langprefs.c:29:
[gettext-bootstrap] /usr/include/dispatch/object.h:143:15: error: expected identifier or '(' before '^' token
[gettext-bootstrap]  typedef void (^dispatch_block_t)(void);
[gettext-bootstrap]                ^
[gettext-bootstrap] /usr/include/dispatch/object.h:362:3: error: unknown type name 'dispatch_block_t'
[gettext-bootstrap]    dispatch_block_t notification_block);
[gettext-bootstrap]    ^
```
E.g. https://github.com/hashdist/hashdist/issues/332 is a duplicate of this bug.

To fix this, we need to ship the fixed header file with our gcc.

To test this issue, just do:
```
$ cat a.c 
#if defined(__APPLE__)
#include <CoreFoundation/CoreFoundation.h>
#endif

int main()
{
    return 0;
}
```
and use system-wide clang:
```
$ clang a.c && ./a.out 
Ok!
```
vs our gcc:
```
$ ~/repos/hashstack/hd_base/bin/gcc a.c && ./a.out 
In file included from /usr/include/dispatch/dispatch.h:51:0,
                 from /System/Library/Frameworks/CoreFoundation.framework/Headers/CFStream.h:15,
                 from /System/Library/Frameworks/CoreFoundation.framework/Headers/CFPropertyList.h:13,
                 from /System/Library/Frameworks/CoreFoundation.framework/Headers/CoreFoundation.h:55,
                 from a.c:2:
/usr/include/dispatch/object.h:143:15: error: expected identifier or ‘(’ before ‘^’ token
 typedef void (^dispatch_block_t)(void);
               ^
/usr/include/dispatch/object.h:362:3: error: unknown type name ‘dispatch_block_t’
   dispatch_block_t notification_block);
   ^
```
The fix is to install the fixed `dispatch/object.h` into the gcc's `$ARTIFACT`, in fact it doesn't even have to be in the profile, you can put it manually into `/Users/certik/.hashdist/bld/gcc/ozg5f7h6k3xv/include/dispatch/object.h` and then things work:
```
$ ~/repos/hashstack/hd_base/bin/gcc a.c && ./a.out 
Ok!
```
To obtain the fixed `object.h`, do:
```
sed 's+typedef void (\^dispatch_block_t)(void)+typedef void* dispatch_block_t+' /usr/include/dispatch/object.h > "object.h"
```

--- a/pkgs/gcc/gcc.yaml
+++ b/pkgs/gcc/gcc.yaml
@@ -27,8 +27,10 @@ build_stages:
     before: configure
     handler: bash
     bash: |
-      mkdir -p "$ARTIFACT/include/dispatch"
-      sed 's+typedef void (\^dispatch_block_t)(void)+typedef void* dispatch_block_t+' /usr/include/dispatch/object.h > "$ARTIFACT/include/dispatch/object.h"
+      if [ -f /usr/include/dispatch/object.h ]; then
+        mkdir -p "$ARTIFACT/include/dispatch"
+        sed 's+typedef void (\^dispatch_block_t)(void)+typedef void* dispatch_block_t+' /usr/include/dispatch/object.h > "$ARTIFACT/include/dispatch/object.h"
+      fi
 
   - name: link_lib64_to_lib
     after: install

--- a/pkgs/gcc/gcc.yaml
+++ b/pkgs/gcc/gcc.yaml
@@ -22,6 +22,14 @@ build_stages:
       patch -up1 < _hashdist/rpath.patch
       sed -i "s|@@ARTIFACT@@|${ARTIFACT}|g" gcc/config/i386/gnu-user.h gcc/config/i386/gnu-user64.h
 
+  - when: platform == 'Darwin'
+    name: fix_object_h
+    before: configure
+    handler: bash
+    bash: |
+      mkdir -p "$ARTIFACT/include/dispatch"
+      sed 's+typedef void (\^dispatch_block_t)(void)+typedef void* dispatch_block_t+' /usr/include/dispatch/object.h > "$ARTIFACT/include/dispatch/object.h"
+
   - name: link_lib64_to_lib
     after: install
     handler: bash

